File: suites/osprey-compute-kernels/reference/cases/listops/listops.hs
{-# LANGUAGE BangPatterns #-}

module Main (main) where

import Control.Monad (replicateM)
import Data.Char (isSpace)
import Data.List (foldl')
import System.IO
  ( IOMode (ReadMode)
  , hGetChar
  , isEOF
  , withBinaryFile
  )
import Text.Read (readMaybe)

minstd :: Int
minstd = 16807

modulus :: Int
modulus = 2147483647

bigMod :: Int
bigMod = 1000000007

r1 :: Int -> Int
r1 x = (x * minstd) `mod` modulus

hashAt :: Int -> Int -> Int
hashAt s i = r1 (r1 (s + i))

readFirstLine :: IO String
readFirstLine = do
  eof <- isEOF
  if eof then pure "" else getLine

readSeed :: IO Int
readSeed = do
  line <- readFirstLine
  let m = maybe 0 id (readMaybe (filter (not . isSpace) line))
  if m == 0
    then pure 1
    else do
      cs <- withBinaryFile "/dev/urandom" ReadMode (replicateM 8 . hGetChar)
      let val = foldl (\acc c -> acc * 256 + fromEnum c) 0 cs
      pure (abs val `mod` 2147483646 + 1)

step :: Int -> Int -> Int
step seed t =
  let s = seed + t * 131
      xs = [hashAt s i `mod` 1000 | i <- [0 .. 3999]]
      sum' = sum xs
      below = length (filter (< 500) xs)
   in sum' + below

main :: IO ()
main = do
  seed <- readSeed
  let acc =
        foldl'
          (\ !a t -> (a + step seed t) `mod` bigMod)
          0
          [0 .. 7]
  print acc

File: suites/osprey-compute-kernels/reference/cases/listops/listops.ml
(* Park-Miller MINSTD RNG *)
let r1 x = (x * 16807) mod 2147483647
let hash_at s i = r1 (r1 (s + i))

let big_mod = 1000000007

let read_seed () =
  let line = try Some (read_line ()) with End_of_file -> None in
  let m =
    match line with
    | Some s -> (match int_of_string_opt (String.trim s) with Some v -> v | None -> 0)
    | None -> 0
  in
  if m = 0 then 1
  else begin
    Random.self_init ();
    (Random.full_int 2147483646) + 1
  end

let () =
  let seed = read_seed () in
  let acc = ref 0 in
  for t = 0 to 7 do
    let s = seed + t * 131 in
    let xs = Array.init 4000 (fun i -> (hash_at s i) mod 1000) in
    let sum = Array.fold_left ( + ) 0 xs in
    let below = Array.fold_left (fun a x -> if x < 500 then a + 1 else a) 0 xs in
    acc := (!acc + sum + below) mod big_mod
  done;
  Printf.printf "%d\n" !acc
